numpy/testing/tests/test_utils.py

```diff
@@ -1086,7 +1086,15 @@ def test_regex(self):
 
 
 def assert_warn_len_equal(mod, n_in_context, py34=None, py37=None):
-    mod_warns = mod.__warningregistry__
+    try:
+        mod_warns = mod.__warningregistry__
+    except AttributeError:
+        # 0 warnings mean the absence of the
+        # of the dictionary attribute, which seems
+        # to have been causing stochastic failures
+        # in parallel test scenarios
```

Comment:
Yeah, OK, now I get it, with parallel tests `__warningsregistry__` was never created, which is OK if there are no warnings. The same could be done in `_get_fresh_mode`, but I think you are right and it is clean here.
The logic is slightly reversed IMO, no `warningregistry` means that no warnings occured.

Will merge in a bit in any case unless someone beats me to it, nice tracking down.

Reply:
Maybe in case it is not quite clear. What happens is that the same module gets reused, the first time the module gets a warning `__warningregistry__` is created. In a non-parallel test run, there is apparently some test that is always run first which makes sure `__warningregistry__` was created (just by creating a warning). In the parallel test run, there is a test that creates no warning and tests for that, so `__warningregistry__` does not exist yet.